### PR TITLE
Cache hit output

### LINKED ISSUE #371: Introduce cache-hit output
**Description:**
I'd like to propose adding a new output called `cache-hit` (like in actions/cache) to show whether dependency cache was hit.

**Justification:**
Since this action utilizes caching mechanic I think it would be needed to have this meta data in output since we can't access internal steps from outside. This would be helpful if, for example, we need add stuff to `PATH` if this cache was hit.

**Are you willing to submit a PR?**
Yes, for sure.

### PR DESCRIPTION
**Pull request by dhvcc**
**Description:**
- Added `cache-hit` output
- Removed useless `await`s in `cache-restore` tests

**Related issue:**
closes #371. Test action run is [here](https://github.com/dhvcc/debug-actions-repo/runs/5802597704?check_suite_focus=true)

**Check list:**
- [x] Mark if documentation changes are required.
- [X] Mark if tests were added or updated to cover the changes.

### CODE REVIEW

**Comment by dhvcc** [comment]
I'm not actually sure if removing awaits belongs in this PR. I also noticed there were no tests for another output - `python-version`. So may be it makes sense to revisit tests and clean them up or add necessary ones?

**Comment by dhvcc** [comment]
Seems like those fails are unrelated to code. I can take a look at the workflows themselves if people are busy

File: src/cache-distributions/cache-distributor.ts [hunk 1]
```diff
@@ -41,12 +42,20 @@ abstract class CacheDistributor {
       restoreKey
     );
 
+    this.handleMatchResult(matchedKey);
+  }
+
+  public handleMatchResult(
+    matchedKey: PromiseReturnType<typeof cache.restoreCache>
```
Comment (dmitry-shibanov) [inline src/cache-distributions/cache-distributor.ts:49]: I think the type should be `string | undefined`.

File: src/cache-distributions/cache-distributor.ts [hunk 2]
```diff
@@ -41,12 +42,20 @@ abstract class CacheDistributor {
...
+  ) {
     if (matchedKey) {
       core.saveState(State.CACHE_MATCHED_KEY, matchedKey);
       core.info(`Cache restored from key: ${matchedKey}`);
     } else {
       core.info(`${this.packageManager} cache is not found`);
     }
+    core.info('cache was hit');
```
Comment (dmitry-shibanov) [inline src/cache-distributions/cache-distributor.ts:57]: Could you please remove this core.info?

File: src/cache-distributions/cache-distributor.ts [hunk 3]
```diff
@@ -41,12 +42,20 @@ abstract class CacheDistributor {
...
+    core.setOutput('cache-hit', Boolean(matchedKey));
```
Comment (dmitry-shibanov) [inline src/cache-distributions/cache-distributor.ts:58]: Could you please adjust logic to check that primary key equals to matchedKey ? We use primary key for pipenv and poetry, but for pip the action has a restore key.

File: src/utils.ts [hunk 4]
```diff
@@ -119,3 +119,12 @@ export function isCacheFeatureAvailable(): boolean {
 
   return true;
 }
+
+// Awaited (typescript 4.5+) polyfill. Not ideal, so use with care
+export type AwaitedPolyfill<T> = T extends PromiseLike<infer U>
```
Comment (dmitry-shibanov) [inline src/utils.ts:124]: I think the `cache-hit` output does not need this type.

File: src/cache-distributions/cache-distributor.ts (hunk 2 above)
Comment (dhvcc) [reply src/cache-distributions/cache-distributor.ts:57]: Oh snap, it was a bit too late hhahaha

File: src/utils.ts (hunk 4 above)
Comment (dhvcc) [reply src/utils.ts:124]: That's for `matchedKey` which comes from `cache.restoreCache`. I think either this or hardcode `string?`. Whatever you prefer really

File: src/cache-distributions/cache-distributor.ts (hunk 1 above)
Comment (dhvcc) [reply src/cache-distributions/cache-distributor.ts:49]: ok cool

File: src/cache-distributions/cache-distributor.ts (hunk 3 above)
Comment (dhvcc) [reply src/cache-distributions/cache-distributor.ts:54]: Is this the check you were interested in? I've updated the PR

Comment (dmitry-shibanov) [reply src/cache-distributions/cache-distributor.ts:54]: Change comparison logic for setOutput `core.setOutput('cache-hit', Boolean(matchedKey))` => `core.setOutput('cache-hit', matchedKey === primaryKey)`

Comment (dhvcc) [reply src/cache-distributions/cache-distributor.ts]: Updated the PR. Doesn't it mean now it's a bit different logic for cache hit output and for  "cache restored from key"?

Comment (dhvcc) [reply src/cache-distributions/cache-distributor.ts]: I get why this comparison is needed now

**Review by brcrista — APPROVED** [review]

**Comment by dmitry-shibanov** [comment]
Hello @dhvcc. I think it should work because the action sets the keys, that is why I consider it will be enough.

**Review by dmitry-shibanov — APPROVED** [review]

**Comment by dmitry-shibanov** [comment]
Hello @dhvcc. I think we'll release it on this week.

**Pull request merged**